File: trunk/acm/spoj4194.py
#!/usr/bin/env python
# @author   FAN Kai (fankai@net.pku.edu.cn), Peking University  
# @date     May 20 22:43:40 CST 2009

from bisect import *
import random

def get_seeds(max):
    roots=(4,7)
    seeds=[]
    for i in range(1, max+1):
        for j in range(0, 2**i):
            n = j
            m = 0
            for k in range(i):
                m *= 10
                m += roots[j % 2]
                j /= 2
            #print m
            seeds.append(m);
    seeds.sort()
    #print seeds
    return seeds

def naive(max):
    seeds = get_seeds(max)
    maxn = 10**max
    size = len(seeds)
    result = seeds
    while True:
        #print size
        result += [a*b for a in result for b in result if a*b<=maxn]
        #print result
        result = list(set(result))
        if len(result) == size: 
            result.sort()
            return result
        size = len(result)

def lucky_number(max):
    seeds = get_seeds(max)
    maxn = 10**max
    result = set()
    v = [1]

    for s in seeds:
        #print len(result)
        #if s * s > maxn: break
        print 's=', s
        a = v
        while a:
            b = []
            for i in a:
                if i*s > maxn: break
                #if i*s in result: continue
                b.append(i*s)
                result.add(i*s)
            #print "a=", a
            print "b=", b
            v += b
            a = b
        #result.sort()
        #print result
        v.sort()
        v = v[:bisect(v, maxn/s)]
        #print v
    result = list(result)
    result.sort()
    return result

def mainflow():
    r = lucky_number(9)
    #print len(r), len(naive(4))
    #print ' '.join(map(str, naive(3)))
    #print len(naive(3)); return
    #print r
    #print ' '.join(map(str, r))
    print len(r);return
    #print naive(3)
    #return 
    n = input()
    for i in range(n):
        #a, b = map(int, raw_input().split())
        #a = random.randint(0, 10**12)
        a = 0
        b = random.randint(a, 10**12)
        print a, b
        print bisect(r, b) - bisect_left(r, a)

mainflow()

File: trunk/acm/cc/spoj4194.cc
/**
 * @file    4194.cc
 * @author  FAN Kai (fankai@net.pku.edu.cn), Peking University 
 * @date    May 21 07:44:28 CST 2009
 */

#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
using namespace std;

vector<int64_t> get_seeds(int power) {
    vector<int64_t> seeds;
    int64_t m = 1;
    int roots[] = {4,7};
    for (int i = 0; i < power; ++i) {
        m *= 2;
        for (int64_t j = 0; j < m; ++j) {
            int64_t a = j, n = 0;
            for (int k = 0; k <= i; ++k) {
                n *= 10;
                n += roots[a%2];
                a /= 2;
            }
            seeds.push_back(n);
        }
    }
    return seeds;
}

set<int64_t> lucky_number(int power) {
    int64_t maxn=1;
    for (int i = 0; i < power; ++i) maxn *= 10;
    vector<int64_t> seeds = get_seeds(power);
    sort(seeds.begin(), seeds.end());
    //cout <<seeds.size() <<endl;
    set<int64_t> lnumbers;
    set<int64_t> all_candidates;
    vector<int64_t> candidates;
    all_candidates.insert(1);
    for (int i = 0; i < seeds.size(); ++i) {
        //cout <<i <<endl;
        //cout <<lnumbers.size() <<" " <<all_candidates.size() <<endl;
        //vector<int64_t>::iterator it = upper_bound(all_candidates.begin(), all_candidates.end(), m);
        //all_candidates.resize(it-all_candidates.begin());
        candidates.clear();
        set<int64_t>::iterator it = all_candidates.begin(); 
        /*
        cout <<"all candidates: ";
        for (set<int64_t>::iterator jt = all_candidates.begin(); 
                jt != all_candidates.end(); ++jt) cout <<*jt <<" "; 
        cout <<endl;
        */
        int64_t maxc = maxn/seeds[i]+1;
        while (it != all_candidates.end()) {
            //cout <<*it <<endl;
            if (*it >=maxc) break;
            candidates.push_back(*it);
            ++it;
        }
        all_candidates.erase(it, all_candidates.end());
        if (0) {
            cout <<"all candidates: ";
            for (set<int64_t>::iterator jt = all_candidates.begin(); 
                    jt != all_candidates.end(); ++jt) cout <<*jt <<" "; 
            cout <<endl;
            cout <<"seed: " <<seeds[i] <<" maxc:" <<maxc<<endl;
        }
        while (candidates.size() > 0) {
            vector<int64_t> results;
            for (int j = 0; j < candidates.size(); ++j) {
                if (candidates[j] > maxc) break;
                int64_t n = seeds[i]*candidates[j];
                if (n > maxn) break;
                lnumbers.insert(n);
                if (n <= maxc) {
                    results.push_back(n);
                    all_candidates.insert(n);
                }
            }
            if (0) {
                cout <<"candidates:" ;
                for (int j = 0; j < candidates.size(); ++j)
                    cout <<candidates[j] <<" ";
                cout <<endl;
                cout <<"results:" ;
                for (int j = 0; j < results.size(); ++j)
                    cout <<results[j] <<" ";
                cout <<endl; 
                cout <<"lnumbers:" <<lnumbers.size() <<endl;
            }
            candidates = results;
        }
        //sort(all_candidates.begin(), all_candidates.end());

    }
    return lnumbers;
}



int main() {
    set<int64_t> vv = lucky_number(9);
    set<int64_t>::iterator it = vv.begin(); 
    vector<int64_t> v;
    for (;it != vv.end(); ++it) v.push_back(*it);
    //for (int i = 0; i < v.size(); ++i) cout <<v[i] <<" ";
    //cout <<v.size() <<endl;

    int n; 
    cin >>n; 
    for (int i = 0; i < n; ++i) {
        int64_t a, b;
        cin >>a >>b;
        cout <<upper_bound(v.begin(), v.end(), b)
            - lower_bound(v.begin(), v.end(), a) <<endl;
    }

    return 0;
}

